pkg/app/api/grpcapi/piped_api.go

```diff
@@ -963,6 +967,93 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
 	return nil, status.Errorf(codes.Unimplemented, "ReportUnregisteredApplicationConfigurations is not implemented yet")
 }
 
+// CreateDeploymentChain creates a new deployment chain object and all required commands to
+// trigger deployment for applications in the chain.
+func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.CreateDeploymentChainRequest) (*pipedservice.CreateDeploymentChainResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	findNodeApps := func(filter *pipedservice.CreateDeploymentChainRequest_ApplicationsFilter) ([]*model.Application, error) {
+		filters := []datastore.ListFilter{
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    projectID,
+			},
+		}
+
+		if filter.AppName != "" {
+			filters = append(filters, datastore.ListFilter{
+				Field:    "Name",
+				Operator: datastore.OperatorEqual,
+				Value:    filter.AppName,
+			})
+		}
+
+		// TODO: Support find node apps by appKind and appLabels.
+
+		apps, _, err := a.applicationStore.ListApplications(ctx, datastore.ListOptions{
+			Filters: filters,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return apps, nil
+	}
+
+	chainNodes := make([]*model.DeploymentChainNode, 0, len(req.Filters))
+	apps := make([]*model.Application, 0)
+	for i, filter := range req.Filters {
+		nodeApps, err := findNodeApps(filter)
+		if err != nil {
+			return nil, err
+		}
+
+		apps = append(apps, nodeApps...)
+		chainNodes = append(chainNodes, &model.DeploymentChainNode{
+			Index:    int32(i + 1),
+			Runnable: false,
+		})
+	}
+
+	dc := model.DeploymentChain{
+		Id:        uuid.New().String(),
+		ProjectId: projectID,
+		Nodes:     chainNodes,
+	}
+
+	// Create a new deployment chain instance to control newly triggered deployment chain.
+	if err := a.deploymentChainStore.AddDeploymentChain(ctx, &dc); err != nil {
+		return nil, err
+	}
+
+	// Make sync application command for applications of the chain.
+	for _, app := range apps {
+		cmd := model.Command{
+			Id:            uuid.New().String(),
+			PipedId:       app.PipedId,
+			ApplicationId: app.Id,
+			ProjectId:     app.ProjectId,
+			Commander:     dc.Id,
+			// TODO: Add new command type in order to separate this in chain sync command with other commands.
+			Type: model.Command_SYNC_APPLICATION,
+			SyncApplication: &model.Command_SyncApplication{
+				ApplicationId: app.Id,
+				SyncStrategy:  model.SyncStrategy_AUTO,
+			},
```

Comment:
Maybe we need a way to distinguish from the normal `SYNC` command emitted via the web to have a better summary for its deployment.
https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/trigger/trigger.go#L259

Reply:
We are on the same page 👍  My plan is to make a new command type (such as `IN_CHAIN_SYNC_APPLICATION`) and filter out all commands with type prefix `IN_CHAIN` and handler it with `OnChainDeterminer` instead of `OnCommandDeterminer`. Also, the new command kind needs to have: deploymentChainId and NodeIndex at least so that the pipedx (who handles the command) will be able to update this deployment chain model its deployment state correctly. That is the thing I'm going to do with that TODO comment.